### PR TITLE
Adding kernel panics table

### PR DESCRIPTION
**Pull request by PoppySeedPlehzr**

### CODE REVIEW

File: osquery/tables/system/darwin/kernel_panics.cpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        }
+      }
+    } else if (boost::starts_with(toks[0], "last loaded kext at") &&
+               toks.size() == 2) {
+      r["last_loaded_kext"] = boost::regex_replace(toks[1], rxSpaces, " ");
+    } else if (boost::starts_with(toks[0], "last unloaded kext at") &&
+               toks.size() == 2) {
+      r["last_unloaded_kext"] = boost::regex_replace(toks[1], rxSpaces, " ");
+    } else if (boost::starts_with(toks[0], "Backtrace") &&
+               std::next(it) != lines.end()) {
```
Comment (theopolis) [inline osquery/tables/system/darwin/kernel_panics.cpp]: Ah, you do need the iterator, cool!

File: osquery/tables/system/darwin/kernel_panics.cpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+      if (registerTokens.size() == 0) {
+        continue;
+      }
+      std::string regLine = "";
+      for (auto& tok_ : registerTokens) {
+        auto regHolder = osquery::split(tok_, ":");
+        if (regHolder.size() != 2) {
+          continue;
+        }
+        auto reg = regHolder[0];
```
Comment (theopolis) [inline osquery/tables/system/darwin/kernel_panics.cpp]: Do you intend to make copies of these values here? You can `auto&` and then `std::move` them into the `regLine` holder.

File: osquery/tables/system/darwin/kernel_panics.cpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    if (timeTokens.size() >= 1 && kDays.count(timeTokens[0]) > 0) {
+      r["time"] = line;
+    }
+
+    if (kRegisters.count(toks[0]) > 0) {
+      auto registerTokens = osquery::split(line, ",");
+      if (registerTokens.size() == 0) {
+        continue;
+      }
+      std::string regLine = "";
```
Comment (theopolis) [inline osquery/tables/system/darwin/kernel_panics.cpp]: You can reduce the scope of `regLine`.

File: osquery/tables/system/darwin/kernel_panics.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+          Row r;
+          readKernelPanic(lf, r);
+          results.push_back(r);
+        }
+      }
+    }
+  };
+
+  if (context.constraints["uid"].notExistsOrMatches("0")) {
+    kernel_panics(kDiagnosticReportsPath);
```
Comment (theopolis) [inline osquery/tables/system/darwin/kernel_panics.cpp]: I don't think you need this, just run the lambda expression logic for every table scan.

File: specs/darwin/kernel_panics.table [hunk 5]
```diff
@@ -0,0 +1,18 @@
+table_name("kernel_panics")
```
Comment (theopolis) [inline specs/darwin/kernel_panics.table]: If you remove the Darwin-specific diction can we use these concepts on Linux later on?

File: osquery/tables/system/darwin/kernel_panics.cpp [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+QueryData genKernelPanics(QueryContext& context) {
+  QueryData results;
+
+  if (context.constraints["uid"].notExistsOrMatches("0")) {
+    std::vector<std::string> files;
+    if (listFilesInDirectory(kDiagnosticReportsPath, files)) {
+      for (const auto& lf : files) {
+        if (alg::ends_with(lf, ".panic")) {
+          Row r;
+          readKernelPanic(lf, r);
```
Comment (theopolis) [inline osquery/tables/system/darwin/kernel_panics.cpp]: This looks odd, why not pass the `QueryResults` as a reference to `readKernelPanic`. Then construct your `Row` within that's methods scope and push back before returning.

File: specs/darwin/kernel_panics.table [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+description("Kernel Panics")
+schema([
+    Column("path", TEXT, "Location of log file"),
+    Column("time", TEXT, "Formatted time of the event"),
+    Column("registers", TEXT, "A space delimited line of register:value pairs"),
+    Column("frame_backtrace", TEXT, "Backtrace of the crashed module"),
+    Column("module_backtrace", TEXT, "Modules appearing in the crashed module's backtrace"),
+    Column("module_dependency", TEXT, "Module dependencies existing in crashed module's backtrace"),
```
Comment (theopolis) [inline specs/darwin/kernel_panics.table]: What about `dependencies`?

File: specs/darwin/kernel_panics.table [hunk 8]
```diff
@@ -0,0 +1,18 @@
+table_name("kernel_panics")
+description("Kernel Panics")
+schema([
+    Column("path", TEXT, "Location of log file"),
+    Column("time", TEXT, "Formatted time of the event"),
+    Column("registers", TEXT, "A space delimited line of register:value pairs"),
+    Column("frame_backtrace", TEXT, "Backtrace of the crashed module"),
+    Column("module_backtrace", TEXT, "Modules appearing in the crashed module's backtrace"),
```
Comment (theopolis) [inline specs/darwin/kernel_panics.table]: Is this just a list of modules? Or is it a trace-style list of frames?

File: specs/darwin/kernel_panics.table [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+    Column("module_dependency", TEXT, "Module dependencies existing in crashed module's backtrace"),
+    Column("name", TEXT, "Process name corresponding to crashed thread"),
+    Column("os_version", TEXT, "Version of the operating system"),
+    Column("kernel_version", TEXT, "Version of the system kernel"),
+    Column("system_model", TEXT, "Physical system model"),
```
Comment (theopolis) [inline specs/darwin/kernel_panics.table]: Can you provide an example of the output here? I'm looking for opportunities to normalize the column names with existing tables like `os_version` and `system_info`.

File: specs/darwin/kernel_panics.table [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+    Column("uptime", BIGINT, "System uptime at kernel panic in nanoseconds"),
+    Column("last_loaded_module", TEXT, "Last loaded module before panic"),
```
Comment (theopolis) [inline specs/darwin/kernel_panics.table]: If you drop the `_module` could this become more platform agnostic, or does the "module" diction hold across Linux and Windows?

File: specs/darwin/kernel_panics.table (hunk 8 above)
Comment (PoppySeedPlehzr) [inline specs/darwin/kernel_panics.table]: It's a trace style list, but there's usually only one listed and the module_dependency follows.

**Comment by facebook-github-bot-2** [comment]
@PoppySeedPlehzr updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2488/files/c09bb30ee99708b0c7f3d48e6171c1e11967f0b0..e1fd046ad2696d05dab08a718818ac39f826ad44)

**Pull request merged**